daemon/logger/loggerutils/rotatefilewriter.go

```diff
@@ -33,45 +44,70 @@ func NewRotateFileWriter(logPath string, capacity int64, maxFiles int) (*RotateF
 	}
 
 	return &RotateFileWriter{
-		f:            log,
-		capacity:     capacity,
-		currentSize:  size,
-		maxFiles:     maxFiles,
-		notifyRotate: pubsub.NewPublisher(0, 1),
+		f:             log,
+		buf:           bytes.NewBuffer(nil),
+		capacity:      capacity,
+		currentSize:   size,
+		lastTimestamp: defaultTimestamp,
+		maxFiles:      maxFiles,
+		notifyRotate:  pubsub.NewPublisher(0, 1),
 	}, nil
 }
 
-//WriteLog write log message to File
-func (w *RotateFileWriter) Write(message []byte) (int, error) {
-	w.mu.Lock()
+//WriteLog write log logEntry to File
+func (w *RotateFileWriter) Write(logEntry *jsonlog.JSONLogs) error {
+	w.writerMu.Lock()
 	if w.closed {
-		w.mu.Unlock()
-		return -1, errors.New("cannot write because the output file was closed")
+		w.writerMu.Unlock()
+		return errors.New("cannot write because the output file was closed")
 	}
+	w.writerMu.Unlock()
+
 	if err := w.checkCapacityAndRotate(); err != nil {
-		w.mu.Unlock()
-		return -1, err
+		return err
 	}
 
-	n, err := w.f.Write(message)
+	w.writerMu.Lock()
+	defer w.writerMu.Unlock()
+	err := logEntry.MarshalJSONBuf(w.buf)
+	if err != nil {
+		return err
+	}
+
+	w.buf.WriteByte('\n')
+	n, err := w.f.Write(w.buf.Bytes())
+	w.buf.Reset()
 	if err == nil {
+		w.lastTimestamp = logEntry.Created
 		w.currentSize += int64(n)
 	}
-	w.mu.Unlock()
-	return n, err
+
+	return err
 }
 
-func (w *RotateFileWriter) checkCapacityAndRotate() error {
+func (w *RotateFileWriter) checkCapacityAndRotate() (err error) {
 	if w.capacity == -1 {
 		return nil
 	}
 
 	if w.currentSize >= w.capacity {
+		w.writerMu.Lock()
+		defer w.writerMu.Unlock()
+
+		ch := make(chan struct{})
+		w.rotateMu.Lock()
+		defer func() {
+			w.rotateMu.Unlock()
+			if err == nil && w.maxFiles > 1 {
+				<-ch
```

Comment:
This is essentially the same as running the compressor in this goroutine, which we don't need to do.
I think what we can do here is something like:


```go
w.rotateMu.Lock()
// stuff
w.rotate(name)
// other stuff

if notCompressing { // whatever this condition needs to be
    w.rotateMu.Unlock()
    return nil
}
go func() {
    compress(name)
    w.rotateMu.Unlock()
}()
return nil
```

This way compression does not block writes but does block rotates.